src/main/java/jenkins/plugins/openstack/compute/slaveopts/BootSource.java

```diff
@@ -208,23 +208,49 @@ protected FormValidation checkNameMatchesOnlyOnce(String value, String endPointU
         private static final long serialVersionUID = -8309975034351235331L;
 
         private final @Nonnull String name;
+        private final @Nonnull Boolean createNewVolume;
+        private final @Nonnull Integer volumeSize;
 
         @DataBoundConstructor
-        public Image(@Nonnull String name) {
+        public Image(@Nonnull String name, @Nonnull Boolean createNewVolume, @Nonnull Integer volumeSize) {
             this.name = name;
+            this.createNewVolume = createNewVolume;
+            this.volumeSize = volumeSize;
         }
 
         @Nonnull
         public String getName() {
             return name;
         }
 
+        @Nonnull
+        public Boolean getCreateNewVolume() {
+            return createNewVolume;
+        }
+
+        @Nonnull
+        public Integer getVolumeSize() {
+            return volumeSize;
+        }
+
         @Override
         public void setServerBootSource(@Nonnull ServerCreateBuilder builder, @Nonnull Openstack os)
                 throws JCloudsCloud.ProvisioningFailedException {
             final List<String> matchingIds = getDescriptor().findMatchingIds(os, name);
             final String id = selectIdFromListAndLogProblems(matchingIds, name, "Images");
-            builder.image(id);
+
+            if (createNewVolume) {
+              final BlockDeviceMappingBuilder volumeBuilder = Builders.blockDeviceMapping()
+                      .sourceType(BDMSourceType.IMAGE)
+                      .destinationType(BDMDestType.VOLUME)
+                      .uuid(id)
```

Comment:
IIUC this takes the configured value from template configuration and uses that as the drive UUID in openstack. Does those not need to be unique, as this will certainly lead to collisions then multiple machines booted from single template? Also, I do not see the need to specify the UUID at all for this use-case.

Reply:
That is not correct. What this does is take the UUID of the image and use that as the source image. 

OpenStack parses the source type and then combines that with the ID. So given the source type is an image and destination is volume, it creates a volume from that image. 

If the ID is not provided, OpenStack won’t be able to create a volume because it wouldn’t know what image to put in it (and well if it’s blank, it won’t boot)

Hope that clears things up

Reply:
Right, makes sense.